src/Mod/Part/Gui/ViewProviderGridExtension.cpp

```diff
@@ -0,0 +1,513 @@
+
+    void getClosestGridPoint(double &x, double &y) const;
+    double getGridSize() const;
+
+    // Configurable parameters (to be configured by specific VP)
+    int GridSizePixelThreshold = 15;
+    int GridNumberSubdivision = 10;
+    int GridLinePattern = 0x0f0f;
+    int GridDivLinePattern = 0xffff;
+    int GridLineWidth = 1;
+    int GridDivLineWidth = 2;
+    unsigned int GridLineColor;
+    unsigned int GridDivLineColor;
+
+    // VP Independent parameters through observer
+    int unitsUserSchema = 0;
+
+    /** Observer for parameter group. */
+    void OnChange(Base::Subject<const char*> &rCaller, const char * sReason) override;
+
+private:
+    void computeGridSize(const Gui::View3DInventorViewer* viewer);
+    void createGrid(bool cameraUpdate = false);
+    void createGridPart(int numberSubdiv, bool divLines, bool subDivLines, int pattern, SoBaseColor* color, int lineWidth = 1);
+
+    bool checkCameraZoomChange(const Gui::View3DInventorViewer* viewer);
+    bool checkCameraTranslationChange(const Gui::View3DInventorViewer* viewer);
+
+    void createEditModeInventorNodes();
+
+    SbVec3f camCenterOnSketch;
+    float camMaxDimension;
+
+private:
+    ViewProviderGridExtension * vp;
+
+    bool enabled = false;
+    double computedGridValue = 10;
+
+    // scenograph
+    SoSeparator * GridRoot;
+};
+
+} // namespace PartGui
+
+
+GridExtensionP::GridExtensionP(ViewProviderGridExtension * vp):
+    camCenterOnSketch(SbVec3f(0., 0., 0.)),
+    camMaxDimension(200.),
+    vp(vp),
+    GridRoot(nullptr)
+{
+    SbColor lineCol(0.7f, 0.7f, 0.7f);
+    GridLineColor = lineCol.getPackedValue();
+    GridDivLineColor = GridLineColor;
+
+    createEditModeInventorNodes();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    unitsUserSchema = hGrp->GetInt("UserSchema", 0); //2 3 5 7 are imperial schemas. 2 3 inches, 5 7 feet
+    hGrp->Attach(this);
+}
+
+GridExtensionP::~GridExtensionP()
+{
+    Gui::coinRemoveAllChildren(GridRoot);
+    GridRoot->unref();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Units");
+    hGrp->Detach(this);
+}
+
+double GridExtensionP::getGridSize() const
+{
+    return computedGridValue;
+}
+
+void GridExtensionP::getClosestGridPoint(double &x, double &y) const
+{
+    auto closestdim = [](double &dim, double gridValue) {
+        dim = dim / gridValue;
+        dim = dim < 0.0 ? ceil(dim - 0.5) : floor(dim + 0.5);
+        dim *= gridValue;
+    };
+
+    closestdim(x, computedGridValue);
+    closestdim(y, computedGridValue);
+
+    //Base::Console().Log("gridvalue=%f, (x,y)=(%f,%f)", computedGridValue, x, y);
+}
+
+bool GridExtensionP::checkCameraZoomChange(const Gui::View3DInventorViewer* viewer)
+{
+    float newCamMaxDimension = viewer->getMaxDimension();
+    if (fabs(newCamMaxDimension - camMaxDimension) > 0) { //ie if user zoomed.
```

Comment:
You can prevent a computation with
```suggestion
    if (newCamMaxDimension - camMaxDimension != 0) { //ie if user zoomed.
```
This said, is it safe to compare a float difference with a strict 0 value ?

Reply:
It may be "safe" in this instance, but even with a set of parenthesis, I think the proposal is difficult to read and understand, given that it is a float.

If this were safe and vital from a performance point of view, then I would suggest encapsulating it in an inline function with a meaningful name.

Reply:
BTW, why just not
```suggestion
    if (newCamMaxDimension != camMaxDimension) { //ie if user zoomed.
```